community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/TriadicBuildPipe.scala

```diff
@@ -19,72 +19,135 @@
                            (val estimatedCardinality: Option[Double] = None)(implicit pipeMonitor: PipeMonitor)
-  extends PipeWithSource(source, pipeMonitor) with RonjaPipe with NoEffectsPipe {
+  extends PipeWithSource(sourcePipe, pipeMonitor) with RonjaPipe with NoEffectsPipe {
 
   override def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState) = {
-    val set = state.triadicSets(triadicSet)
     input.filter {
       ctx =>
-        ctx(identifier) match {
-          case n: Node => !set.contains(n.getId)
+        ctx(target) match {
+          case n: Node =>
+            !state.triadicState(seen).contains(n.getId)
           case _ => false
         }
     }
   }
 
   override def planDescriptionWithoutCardinality: InternalPlanDescription =
-    PlanDescriptionImpl(this.id, "TriadicProbe", SingleChild(source.planDescription),
-      Seq(KeyNames(Seq(triadicSet, identifier))), identifiers)
+    PlanDescriptionImpl(this.id, "TriadicProbe", SingleChild(sourcePipe.planDescription),
+      Seq(KeyNames(Seq(seen, target))), identifiers)
 
   override def withEstimatedCardinality(estimated: Double) =
     copy()(Some(estimated))
 
   override def symbols =
-    source.symbols
+    sourcePipe.symbols
 
   override def dup(sources: List[Pipe]) = {
     val (head :: Nil) = sources
-    copy(source = head)(estimatedCardinality)
+    copy(sourcePipe = head)(estimatedCardinality)
   }
 }
+
+abstract class LazyGroupingIterator[ROW >: Null <: AnyRef](val input: Iterator[ROW]) extends AbstractIterator[ROW] {
+  def setState(state: PrimitiveLongSet)
+  def getKey(row: ROW): Any
+  def getValue(row: ROW): Option[Long]
+
+  var current: Iterator[ROW] = null
+  var nextRow: ROW = null
+
+  override def next() = if(hasNext) current.next() else Iterator.empty.next()
+
+  override def hasNext: Boolean = {
+    if (current != null && current.hasNext)
+      true
+    else {
+      val firstRow = if(nextRow != null) {
+        val row = nextRow
+        nextRow = null
+        row
+      } else if(input.hasNext) {
+        input.next()
+      } else null
+      if (firstRow == null) {
+        current = null
+        setState(null)
+        false
+      }
+      else {
+        val buffer = new ListBuffer[ROW]
+        val valueSet = Primitive.longSet()
+        setState(valueSet)
+        buffer += firstRow
+        update(valueSet, firstRow)
+        val key = getKey(firstRow)
+        // N.B. should we rewrite takeWhile to a while-loop?
+        buffer ++= input.takeWhile{ row =>
+          val s = getKey(row)
+          if (s == key) {
+            update(valueSet, row)
+            true
+          } else {
+            nextRow = row
+            false
+          }
+        }
+        current = buffer.iterator
+        true
+      }
+    }
+  }
+
+  def update(triadicSet: PrimitiveLongSet, row: ROW): AnyVal = {
+    for (value <- getValue(row))
+      triadicSet.add(value)
+  }
+}
```

Comment:
github thinks there should be a new line at the end-of-file, I thought we had some style check that enforces that as well. Obviously not important though.